Bump euclid to 0.14.x.

nical opened the pull request:
- [x] `./mach build -d` does not report any errors (kinda, need webrender published and Cargo.toml fixed up)
- [x] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

<!-- Either: -->
- [ ] There are tests for these changes OR
- [x] These changes do not require tests because it is a refactoring in which the difference is mostly a compile-time/strong-typing thing with no change to the logic.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/17184)
<!-- Reviewable:end -->

nical committed:
Bump euclid to 0.14.

nical committed:
Make BaseFlow::stacking_relative_position a vector.

highfive commented:
Thanks for the pull request, and welcome! The Servo team is excited to review your changes, and you should hear from @asajeffrey (or someone else) soon.

highfive commented:
Heads up! This PR modifies the following files:
 * @bholley: components/style/gecko_bindings/sugar/ns_timing_function.rs, components/style/properties/gecko.mako.rs, components/style/animation.rs, components/style/properties/helpers/animated_properties.mako.rs, components/style/properties/properties.mako.rs and 7 more
 * @jgraham: components/webdriver_server/Cargo.toml
 * @wafflespeanut: python/tidy/servo_tidy_tests/rust_tidy.rs
 * @emilio: components/style/gecko_bindings/sugar/ns_timing_function.rs, components/style/properties/gecko.mako.rs, components/style/animation.rs, components/layout/multicol.rs, components/layout/list_item.rs and 21 more
 * @fitzgen: components/script/dom/htmlcanvaselement.rs, components/script/timers.rs, components/script/dom/imagedata.rs, components/script_traits/script_msg.rs, components/script_traits/script_msg.rs and 25 more
 * @kichjang: components/script/dom/htmlcanvaselement.rs, components/script/timers.rs, components/script/dom/imagedata.rs, components/script_traits/script_msg.rs, components/script_traits/script_msg.rs and 25 more
 * @asajeffrey: components/constellation/browsingcontext.rs, components/constellation/constellation.rs, components/webdriver_server/Cargo.toml, components/constellation/pipeline.rs, components/constellation/Cargo.toml

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!

nical commented:
This requires webrender the equivalent WebRender and rust-azure changes to be merged/published, in order to build (I did get it to work locally), but the review is going to be a lot of work so I am submitting this early for people to start looking at it. This pull request is enormous, and I am truly sorry for this.
We made some major changes to euclid, in particular the separation of points and vector types. See https://github.com/servo/euclid/pull/159.

cc @nox

bors-servo commented:
:umbrella: The latest upstream changes (presumably #17186) made this pull request unmergeable. Please resolve the merge conflicts.

asajeffrey commented:
Gosh, you're not kidding about this being a big change!

Are most of the changes to euclid naming? Would it be possible/useful to publish an intermediate version of euclid with both naming schemes?

nical commented:
The biggest change is the separation between vector and points. The rest of the diff is trivial changes like renaming Matrix2D/4D into Transform2D/3D and cleaning up the namespaces (`use euclid::rect::Rect;` becomes `use euclid::Rect`).
There is also two commits about SVG and animations that are not actually from this PR but that appear in the diff. After I have rebased this, they will disappear from the diff.

Once I have rebased this patch though, most of the non-trivial changes won't be in the diff, and if you want I can roll back the the Matrix/Transform renaming (this specific part doesn't occupy much room in the diff though) since I left an alias for the old name in euclid marked as deprecated, so it can be postponed if servo build does not deny warnings.

For the other changes, I am sorry but I spent way too much time getting this euclid bump into all of the servo crates outside the main repo, getting reviews an getting things published on crates in the correct order. I ended up doing mostly that for more than a week and I won't do it again. Had I known how much work this would be, I would not have made these changes to euclid.

nical commented:
deprecation warnings). The total diff is a lot less scary to look at now.
Let me know if you prefer things squashed as is and move the matrix/transform stuff to a separate PR or if you want it in this one.

I kept another commit separated: the one that turns `BaseFlow::stacking_relative_position` into a vector instead of a point. For this one I wasn't 100% sure whether it made most sense as a vector or a point (short of knowing the layout code well enough), so it is still in its own commit which can be removed easily if need be.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #16752) made this pull request unmergeable. Please resolve the merge conflicts.

mrobinson commented on Cargo.toml:
```diff
@@ -19,3 +19,7 @@ opt-level = 3
+
+"https://github.com/servo/webrender#0.40.0" = { path = "../webrender/webrender/" }
+"https://github.com/servo/webrender#webrender_traits:0.40.0" = { path = "../webrender/webrender_traits/" }
+"https://github.com/servo/rust-azure#azure:0.16.0" = { path = "../rust-azure/" }
```
Now that azure has the WebRender bump as well, I guess this can be removed.

asajeffrey commented on components/layout/fragment.rs:
```diff
@@ -2861,11 +2861,11 @@ impl Fragment {
-                                                         transform_origin_y,
-                                                         transform_origin_z);
+                                                            transform_origin_y,
+                                                            transform_origin_z);
```
Nit: I think our indentation style is the old version of the code.

asajeffrey commented on components/layout/fragment.rs:
```diff
@@ -3099,9 +3099,9 @@ impl Overflow {
-    pub fn translate(&mut self, point: &Point2D<Au>) {
-        self.scroll = self.scroll.translate(point);
-        self.paint = self.paint.translate(point);
+    pub fn translate(&mut self, by: &Vector2D<Au>) {
```
This makes a lot more sense with the new type name!

asajeffrey commented on components/layout/webrender_helpers.rs:
```diff
@@ -86,12 +86,22 @@ trait ToPointF {
 }
 
+trait ToVectorF {
+    fn to_vectorf(&self) -> webrender_traits::LayoutVector2D;
```
Nit: the function name is not very self-explanatory.

asajeffrey reviewed (commented):
Looks perfectly sensible, and a lot of work! Just a couple of nits.

SimonSapin commented:
Looks good to me.

---

Reviewed 39 of 87 files at r1, 41 of 51 files at r4, 9 of 9 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[components/canvas/canvas_paint_thread.rs, line 651 at r5](https://reviewable.io:443/reviews/servo/servo/17184#-KmWR55uRM35K1KZnz6X:-KmWR55uRM35K1KZnz6Y:bkww5cy) ([raw file](https://github.com/servo/servo/blob/cee0715de90f86cae0ff98f73a6724bfbf045abf/components/canvas/canvas_paint_thread.rs#L651)):*
> ```Rust
> 
>         // Step 6.
>         let dest_rect = dirty_rect.translate(&offset.to_vector()).to_i32();
> ```

Should `offset` be a vector to start with?

---

*[components/script/dom/document.rs, line 869 at r5](https://reviewable.io:443/reviews/servo/servo/17184#-KmWQwUTl8NwFar37oeQ:-KmWQwUTl8NwFar37oeR:b-cxdltz) ([raw file](https://github.com/servo/servo/blob/cee0715de90f86cae0ff98f73a6724bfbf045abf/components/script/dom/document.rs#L869)):*
> ```Rust
>                 let rect = iframe.upcast::<Element>().GetBoundingClientRect();
>                 let child_origin = Point2D::new(rect.X() as f32, rect.Y() as f32);
>                 let child_point = client_point - child_origin.to_vector();
> ```

Could `child_origin` be a vector to start with? (Here and 3 others in this file.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/17184#-:-KmWSpCNVaojlpJzmTuG:b9j9n43)*
<!-- Sent from Reviewable.io -->

nical commented:
The review comments have been addressed, should we land this now?

SimonSapin commented:
Thank you!

@bors-servo r+

---

Reviewed 3 of 13 files at r6, 9 of 9 files at r7, 1 of 10 files at r8, 9 of 9 files at r9.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/17184#-:-KmakHE8gEg37y6ZxOHm:b-cr0ec)*
<!-- Sent from Reviewable.io -->

bors-servo commented:
:pushpin: Commit 87e6a6c has been approved by `SimonSapin`

<!-- @bors-servo r=SimonSapin 87e6a6c855c6b30741c223447553deeaf5c30b9b -->

bors-servo commented:
:hourglass: Testing commit 87e6a6c855c6b30741c223447553deeaf5c30b9b with merge 4e25084c171f47d7ca1400ed133bf576fe122e07...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/7825)

SimonSapin commented:
@bors-servo r+

---

Reviewed 15 of 24 files at r10, 9 of 9 files at r11.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/17184#-:-Kmay8YlSCsgRvndNCsi:bkg9nno)*
<!-- Sent from Reviewable.io -->

bors-servo commented:
:pushpin: Commit 997608f has been approved by `SimonSapin`

<!-- @bors-servo r=SimonSapin 997608f11f6dfa79291ead25cae46a0538f2a3dc -->

bors-servo commented:
:hourglass: Testing commit 997608f11f6dfa79291ead25cae46a0538f2a3dc with merge 18653f69581693a5bae1ce4e350e78bc16159b08...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/7484), [arm32](http://build.servo.org/builders/arm32/builds/7698), [arm64](http://build.servo.org/builders/arm64/builds/7738), [linux-dev](http://build.servo.org/builders/linux-dev/builds/7827), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/4582), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/4606), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/7809), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/294), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/293), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/4512), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/4579), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/287), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/291), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/3054)
Approved by: SimonSapin
Pushing 18653f69581693a5bae1ce4e350e78bc16159b08 to master...
